modules/transaction-payment/src/lib.rs

```diff
@@ -611,59 +673,206 @@ where
+				if treasury_balance < SwapBalanceThreshold::<T>::get(supply_currency_id) {
+					let trading_path = Self::get_trading_path_by_currency(&treasury_account, supply_currency_id);
+					if let Some(trading_path) = trading_path {
+						// using all supply asset to swap out native asset, no need to consider ED limitation here.
+						let treasury_supply_balance =
+							T::MultiCurrency::free_balance(supply_currency_id, &treasury_account);
+						if let Ok(swap_native_balance) =
+							T::DEX::swap_with_exact_supply(&treasury_account, &trading_path, treasury_supply_balance, 0)
 						{
-							// successfully swap, break iteration
-							break;
+							// calculate and update new rate of supply asset relative to native asset
+							let new_native_balance =
+								rate.saturating_mul_int(swap_native_balance.saturating_add(treasury_balance));
+							rate = Ratio::saturating_from_rational(new_native_balance, PoolSize::<T>::get());
+							TokenFixedRate::<T>::insert(supply_currency_id, rate);
 						}
 					}
-					_ => {}
+				}
+
+				// use fix rate to calculate the amount of supply asset that equal to native asset.
+				// exchange user's supply asset and treasury's native asset.
+				let token_amount: u128 = rate.saturating_mul_int(amount);
+				T::MultiCurrency::transfer(supply_currency_id, who, &treasury_account, token_amount)?;
+				T::Currency::transfer(&treasury_account, who, amount, ExistenceRequirement::KeepAlive)?;
+				Ok(())
+			}
+		}
+	}
+
+	/// Get trading path by user.
+	fn get_trading_path(who: &T::AccountId) -> Vec<Vec<CurrencyId>> {
+		let default_fee_swap_path_list = T::DefaultFeeSwapPathList::get();
+		if let Some(trading_path) = AlternativeFeeSwapPath::<T>::get(who) {
+			vec![vec![trading_path.into_inner()], default_fee_swap_path_list].concat()
+		} else {
+			default_fee_swap_path_list
+		}
+	}
+
+	/// Get trading path by user and supply asset.
+	fn get_trading_path_by_currency(who: &T::AccountId, supply_currency_id: CurrencyId) -> Option<Vec<CurrencyId>> {
+		let fee_swap_path_list: Vec<Vec<CurrencyId>> = Self::get_trading_path(who);
+		for trading_path in fee_swap_path_list {
+			if let Some(currency) = trading_path.first() {
+				if *currency == supply_currency_id {
+					return Some(trading_path);
+				}
+			}
+		}
+		None
+	}
+
+	/// The sub account derivated by `TreasuryPalletId`.
+	fn treasury_sub_account_id(id: CurrencyId) -> T::AccountId {
+		T::TreasuryPalletId::get().into_sub_account(id)
+	}
+
+	/// Setting some initial value when on_runtime_upgrade.
+	pub fn on_runtime_upgrade(balance: Balance, currency_id: CurrencyId, rate: Ratio) -> DispatchResult {
+		let account = Self::treasury_sub_account_id(currency_id);
+		T::Currency::transfer(
+			&T::TreasuryAccount::get(),
+			&account,
+			balance,
+			ExistenceRequirement::KeepAlive,
+		)?;
+		SwapBalanceThreshold::<T>::insert(currency_id, balance / 5);
+		TokenFixedRate::<T>::insert(currency_id, rate);
+		PoolSize::<T>::set(balance);
```

Comment:
this needs to be different per currency id? otherwise it shouldn't be updated here